### PR TITLE
A higher level language for schedule scripts

### PR DESCRIPTION
**Pull request by AngheloAlf**
Inspired by the massive schedule script of Anju I decided that we need a better way to handle the schedule scripts.

A few reasons why I don't like the current way we handle them:
- They can be huge and with complex control flows that are hard to follow.
- Very hard to modify. They have branch offsets hardcoded into the command themselves, so adding or removing commands means updating all the offsets on the schedule script.
- Writing a new script from scratch is terrible, mainly due to the same reasons of trying to modify one.

Taking all of this into consideration, I made a higher level language that gets compiled into the macros we currently have. It has a syntax inspired by C, but with many distinctions like not requiring colons, using commands instead of normal if statements, very limited operators, not having variables, etc.

I documented this new language as extensively as I could on the `docs/schedule_scripting_language.md` document. Let me know if I missed something or something is not clear. I'm also very open to suggestions and changes to the language itself.

I converted all the scripts to this new language, with the only exception of Anju's script because it is currently in PR.

The compiler is at `tools/buildtools/schc.py`. I didn't try to be very clever with it, just made it good enough to produce matching output. In the future somebody can take it and improve it by adding new operators or actual optimization passes.

### CODE REVIEW

File: tools/buildtools/schc.py [hunk 1]
```diff
@@ -0,0 +1,923 @@
...
+    # Extra tokens
+    TokenType.ELSE:                 TokenProperties(isExtraToken=True),
+    TokenType.BRACE_OPEN:           TokenProperties(isExtraToken=True),
+    TokenType.BRACE_CLOSE:          TokenProperties(isExtraToken=True),
+    TokenType.ARGS:                 TokenProperties(isExtraToken=True),
+    TokenType.NOT:                  TokenProperties(isExtraToken=True),
+    TokenType.LABEL:                TokenProperties(isExtraToken=True),
+}
+
+tokenLiterals: dict[str, TokenType] = {x.value: x for x in TokenType}
```
Comment (Dragorn421) [inline tools/buildtools/schc.py:180]: iirc e.g. `TokenType("{")` (not `TokenType["{"]`) will be the same as `tokenLiterals["{"]`

File: tools/buildtools/schc.py [hunk 2]
```diff
@@ -0,0 +1,923 @@
...
+        if self.index >= len(self.tokens):
+            return None
+        token = self.tokens[self.index]
+        self.index += 1
+        return token
+
+    def unget(self) -> None:
+        self.index -= 1
+        if self.index < 0:
+            raise RuntimeError()
```
Comment (Dragorn421) [inline tools/buildtools/schc.py:232]:
```suggestion
            raise Exception("bad programming, can't unget")
```
:shrug: on the message but I do think you should have one, and also use Exception, or make a `class SchcError(Exception): pass`

Also I think the check should come before decrementing index so you don't leave the class in a bad state

File: tools/buildtools/schc.py [hunk 3]
```diff
@@ -0,0 +1,923 @@
...
+    def remainingTokens(self) -> int:
+        return len(self.tokens) - self.index
+
+    def reset(self) -> None:
+        self.index = 0
+
+
+# Strips comments
+def preprocess(contents: str, filename: str) -> str:
+    result = ""
```
Comment (Dragorn421) [inline tools/buildtools/schc.py:243]: ftr (not an issue here because the schedule scripts will never be big) appending characters one at a time to a string in python is quadratic complexity, i.e. if `s` is of length N>>1 then
```py
r = ""
for c in s:
  r += c
```
takes a time proportional to N^2

Because the string is reallocated and copied every loop. CPython can sometimes optimize this but YMMV

What you want to do in python is build a list of characters and concatenate all at once, it's typically unexpectedly faster
```py
L = []
for c in s:
  L.append(c)
r = "".join(L)
```

Again in this case it's fine, you're processing less than, idk, ~1000 characters per string ever

File: tools/buildtools/schc.py [hunk 4]
```diff
@@ -0,0 +1,923 @@
...
+@dataclasses.dataclass
+class LabeledExpression:
+    # Index relative to the parent list
+    index: int
+
+    token: Token
+    args: Token|None
+
+    # Label to jump into this command
+    labelName: str
```
Comment (Dragorn421) [inline tools/buildtools/schc.py:613]: Btw I'm picking this line randomly but why not use the *de facto* standard for python comments on fields and variables,

```suggestion
    labelName: str
    """Label to jump into this command"""
```
iirc

File: tools/buildtools/schc.py [hunk 5]
```diff
@@ -0,0 +1,923 @@
...
+            # Keep the labelname but ignore the label itself
+            labelName = token.tokenLiteral
+            continue
+
+        tokenProperties = token.getProperties()
+
+        left = expr.left
+        right = expr.right
+
+        # Some commands need to invert theirs `if_`/`else` bodies to match the corresponding command
```
Comment (Dragorn421) [inline tools/buildtools/schc.py:650]:
```suggestion
        # Some commands need to invert their `if_`/`else` bodies to match the corresponding command
```

File: tools/buildtools/schc.py [hunk 6]
```diff
@@ -0,0 +1,923 @@
...
+            left, right = right, left
+
+        # Invert the bodies if the expression was negated
+        if expr.negated:
+            left, right = right, left
+
+        currentIndex = index
+        index += 1
+
+        # Linealize the left body
```
Comment (Dragorn421) [inline tools/buildtools/schc.py:661]:
```suggestion
        # Linearize the left body
```

File: tools/buildtools/schc.py [hunk 7]
```diff
@@ -0,0 +1,923 @@
...
+                fatalError(f"Branch command without arguments?", branchExpr.token.filename, branchExpr.token.lineNumber, branchExpr.token.columnNumber)
+            targetLabel = branchExpr.args.tokenLiteral.strip()
+        else:
+            sub, index = convertTreeIntoLabeledList(right, index)
+            subResults += sub
+            if len(sub) != 0:
+                # Get the label name of the first expression
+                targetLabel = sub[0].labelName
+
+        # If there's no user-defined label, then autogenerate one. Use dots to avoid name crashes with user-defined ones
```
Comment (Dragorn421) [inline tools/buildtools/schc.py:680]:
```suggestion
        # If there's no user-defined label, then autogenerate one. Use dots to avoid name clashes with user-defined ones
```
The japanengrish is getting to you monkaS

File: tools/buildtools/schc.py [hunk 8]
```diff
@@ -0,0 +1,923 @@
...
+
+    return result
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Compiler for the high level schedule language", prog=__prog_name__)
+
+    parser.add_argument("input", help="Schedule script path", type=Path)
+    parser.add_argument("-o", "--output", help="Output path. Will print to stdout if omitted", type=Path)
+    parser.add_argument("-g", type=int, nargs="?", const=1, default=0, dest="debuggingLevel", metavar="level", help="Emit debugging information on the generated macros. Level 0 means no debugging information. Passing no level at all implies level 1. Defaults to level 0")
```
Comment (Dragorn421) [inline tools/buildtools/schc.py:847]:
```suggestion
    parser.add_argument("-g", type=int, nargs="?", const=1, default=0, dest="debuggingLevel", metavar="level", help="Emit debugging information on the generated macros. Level 0 means no debugging information. -g is like -g1. Default is -g0")
```
Is a lot less confusing to me if I got it right

**Review by Dragorn421 — COMMENTED** [review]
Didn't know this was a thing, interesting system
Great job!

File: docs/schedule_scripting_language.md [hunk 9]
```diff
@@ -0,0 +1,901 @@
...
+
+##### `if_before_time` non-generics
+
+`if_before_time_s` and `if_before_time_l`
+
+##### `if_before_time` notes
+
+This command performs the contrary check to `if_since_time`.
+
+#### `if_since_time`
```
[the remark this answers is not in the document]
Comment (AngheloAlf) [reply docs/schedule_scripting_language.md:467]: To me "after" sounds like a `>` check, while "since" sounds like a `>=` check.
At least to me it felt like an important distinction to make since this command is just syntactic sugar for `not if_before_time` (like, it isn't a command that maps exactly to a command from `z_schedule.c`).
If this difference isn't worth the trouble then we could just go with `if_after_time`

File: docs/schedule_scripting_language.md [hunk 10]
```diff
@@ -0,0 +1,901 @@
...
+
+### Operators
+
+Operators can be applied to some commands.
+
+Currently only one operator is allowed on the language, the [`not`](#not) operator.
+
+#### `not`
+
+A label is a special kind of command that doesn't get compiled into the actual
```
[the remark this answers is not in the document]
Comment (AngheloAlf) [reply docs/schedule_scripting_language.md:694]: whoops. bad copypaste :sweat_smile:

File: docs/schedule_scripting_language.md [hunk 11]
```diff
@@ -0,0 +1,901 @@
...
+}
+```
+
+##### `if_since_time` non-generics
+
+`if_since_time_s` and `if_since_time_l`
+
+##### `if_since_time` notes
+
+This command performs the contrary check to `if_before_time`.
```
[the remark this answers is not in the document]
Comment (AngheloAlf) [reply docs/schedule_scripting_language.md:492]: I made it this way in case somebody checked the documentation of one command but not the other one

File: docs/schedule_scripting_language.md [hunk 12]
```diff
@@ -0,0 +1,901 @@
...
+
+```c
+if_week_event_reg (WEEKEVENTREG_HAD_MIDNIGHT_MEETING) {
+    return_none
+}
+```
+
+#### `return_empty`
+
+The schedule finished without changing the previous value.
```
[the remark this answers is not in the document]
Comment (AngheloAlf) [reply docs/schedule_scripting_language.md:632]: I agree, but it is hard to come up with a good name when the only difference between them is this:
https://github.com/zeldaret/mm/blob/10a14feb2c9a484597fce6e29df0ab5f7fa62848/src/code/z_schedule.c#L93-L103

File: docs/schedule_scripting_language.md [hunk 13]
```diff
@@ -0,0 +1,901 @@
...
+The schedule script is run until a return command is executed. There are various
+return commands that allow different things:
+
+- [Return none](#return_none): The schedule finished without returning a value.
+- [Return empty](#return_empty): The schedule finished without changing the
+  previous value.
+- [Return s](#return_s): The script returns a value that's 1 byte long.
+- [Return l](#return_l): The script returns a value that's 2 bytes long (Note
+  this is bugged and will be truncated to 1 byte).
+- [Return time](#return_time): Returns a time range and a 1 byte value.
```
Comment (isghj5) [inline docs/schedule_scripting_language.md:109]: Is this 1 extra byte actor specific?

File: docs/schedule_scripting_language.md [hunk 14]
```diff
@@ -0,0 +1,901 @@
...
+- Argument 1: Minute component of the start time
+- Argument 2: Hour component of the end time
+- Argument 3: Minute component of the end time
+
+##### `if_time_range` example
+
+```c
+// Checks if the current time is between 18:00 ~ 6:00
+if_time_range (18, 0, 6, 0) {
+    return_time (18, 0, 6, 0, 2)
```
Comment (isghj5) [inline docs/schedule_scripting_language.md:367]: Does supporting typical time syntax directly in the macro adds a lot of complexity?
like
```
if_time_range (18:00, 6:00) {
    return_time (18:00, 6:00, 2)
```
I'm sure there is some matching reason to keep the raw data format, but if possible I would like to consider readability here, even if it's just a redundant command.

File: docs/schedule_scripting_language.md [hunk 15]
```diff
@@ -0,0 +1,901 @@
...
+    }
+} else {
+    if_time_range (10, 0, 20, 0) {
+        branch (label_0x8)
+    } else {
+        label_0xF:
+        return_s (21)
+    }
+}
+```
```
Comment (isghj5) [inline docs/schedule_scripting_language.md:525]: ... the game actually does this in schedule format? its not just a bad match in the decompiled code, or is this just an example?

File: tools/buildtools/schc.py (hunk 3 above)
Comment (AngheloAlf) [reply tools/buildtools/schc.py:243]: Compiling all the scripts with no changes to the compiler
```
real    0m4.517s
user    0m3.518s
sys     0m0.852s
```
With this suggestion:
```
real    0m4.437s
user    0m3.436s
sys     0m0.927s
```

Since the files are so small seems to not really matter much. But thanks for pointing it out, this can help in other tools of mine like spimdisasm lol

File: docs/schedule_scripting_language.md (hunk 13 above)
Comment (AngheloAlf) [reply docs/schedule_scripting_language.md:109]: Yes, it serves the same purpose as the returned value of `return_s`/`return_l`, but adding a time range.

Comment (AngheloAlf) [reply docs/schedule_scripting_language.md:109]: For your second question, the compiler is _very_ dumb, it doesn't even try to check for arguments being correct. It just assumes everything between parentheses is correct and passes them to the generated macros.
I made it this way to not have to worry to support about any symbol/macro/enum defined in C, just past it through and let the C compiler to handle them

File: docs/schedule_scripting_language.md (hunk 15 above)
Comment (AngheloAlf) [reply docs/schedule_scripting_language.md:525]: This is an example taken verbatim from `src/overlays/actors/ovl_En_Door/D_808676F0.schedule`.

I converted all commands almost literally from their macro versions to this new language, and this script literally does this. You can see it yourself by following the control flow on the EnDoor actor:
https://github.com/zeldaret/mm/blob/10a14feb2c9a484597fce6e29df0ab5f7fa62848/src/overlays/actors/ovl_En_Door/z_en_door.c#L179-L185

As for how this script ended up having a control flow so weird, idk. Maybe they upgraded their tooling to have optimization passes (a lot of schedules don't seem to be optimized at all) or maybe the EnDoor dev was very passionate about optimizing door's schedules

File: docs/schedule_scripting_language.md (hunk 14 above)
Comment (AngheloAlf) [reply docs/schedule_scripting_language.md:367]: As mentioned in the other comment, the compiler doesn't process the arguments at all, they are passed verbatim to the generated macros, so adding something like this would indeed add a bit of complexity, but not that much I suppose. I would need to look into it

**Review by hensldm — COMMENTED** [review]
This is really nice. I have some thoughts which tbh I'm not sure are good or not.

1. Instead of different files for each schedule script, just use one file and separate them by name (maybe function like syntax). The compiler would than need to output the full array, and just use the name as the variable name.
2. Not related to the compiler per say, but thoughts on a `scehduleScript` typedef? (Tharo has a MsgEventScript one in his msgevent branch and I really like it)
3. Perhaps we shorten the file extension name to .schc (I would say .sch or .schd but both of those are actual file extensions for other things) to match the python script name?

**Comment by isghj5** [comment]
> Instead of different files for each schedule script, just use one file and separate them by name (maybe function like syntax). The compiler would than need to output the full array, and just use the name as the variable name.
I feel like this whole project has been moving away from putting everything in one file, though. Separating the `code` blob into organizable separate files, splitting z64.h into many smaller organized headers, moving large raw assets in overlays to their own asset file out of the c file,  someone was even advocating the possibility of splitting `object` assets into different files at one point.

**Comment by hensldm** [comment]
> > Instead of different files for each schedule script, just use one file and separate them by name (maybe function like syntax). The compiler would than need to output the full array, and just use the name as the variable name.
> 
> I feel like this whole project has been moving away from putting everything in one file, though. Separating the `code` blob into organizable separate files, splitting z64.h into many smaller organized headers, moving large raw assets in overlays to their own asset file out of the c file, someone was even advocating the possibility of splitting `object` assets into different files at one point.

This is true, but having each individual schedule script being its own file just seems like too much (at least imo) especially when like EnDoor quite a few are just `return_none` and nothing else. If we also do MsgEvent scripts something like this (though not sure if it is possible yet) folders are going to get pretty bloated with a lot of small files.

**Comment by AngheloAlf** [comment]
I think we should try to make some kind of balance on how much everything is splitted.

I originally made this language with 1 script per file because it was simpler and it was what most actors do. Also I kinda like how each array is defined in the C file and not just everything included like with some overlays that include assets. But I can also see the annoyance of having so many. I guess I would need to work on it to see how it ends up looking so we could decide

When MsgEvent gets merged I would definitely be interested on looking if an equivalent language for this is possible too.

Having an `ScheduleScript` typedef could be interesting yeah, I'll add it.

I went with `.schedule` to avoid clashing with existing extensions. `schc` would be a bit unfortunate since it is short for "schedule compiler", but I guess we could just use it

**Comment by hensldm** [comment]
> I went with `.schedule` to avoid clashing with existing extensions. `schc` would be a bit unfortunate since it is short for "schedule compiler", but I guess we could just use it

Ah that makes sense. `.schedule` works than.

**Review by isghj5 — APPROVED** [review]
Want to clarify, even though I was arguing one way above I don't really have a strong preference between separating per-schedule or lumping all together into one file, I can approve this as-is because its a huge improvement, and we change organization later if there is a need.

**Comment by AngheloAlf** [comment]
I want to mention that I have been procrastinating on trying to implement the multiple scripts in a file thingy, shorry :sweat_smile:

**Comment by AngheloAlf** [comment]
Finally implemented the requested multiple schedule scripts per file

**Review by hensldm — APPROVED** [review]
Thanks for adding the feature :) . Personally I think it is an improvement, but would still be interested in other's thoughts.

**Comment by yemenalcyone230** [comment]
.pleh deen i dna kcuts fo tros m'I .ti fo dir teg ot kcolb eht no enihs ot esu nac I taht moor taht ni thgil on ereht ,lleW.roolf tsrif eht no s'tI ?secnad naissuR gniod sdaeder 3 sah taht si kcolb nus egral eht erehw eno eht ,detacol si retsamroolf eht erehw moor eht wonk uoY

**Comment by altoidmerope294** [comment]
krow t now ti dna lios eht raen nottub eht gniyrt neeb ev I

**Comment by ginomeelectra202** [comment]
.pleh esaelP.selkcerf dna riah neerg eht htiw ydal/tsinoitpecer eht fo eman eht wonk enoyna seoD

File: docs/schedule_scripting_language.md [hunk 16]
```diff
@@ -0,0 +1,925 @@
...
+instead.
+
+A long command uses two bytes to store the branch distance, meaning the command
+itself will be one byte longer.
+
+There are no commands that use more than 2 bytes to store the branch distance.
+
+The short and long distinction also exists for the returned value, in case the
+user wants to return a value that wouldn't fit on a single unsigned byte and
+requires an unsgined short (two bytes) instead. Please note that the vanilla
```
Comment (engineer124) [inline docs/schedule_scripting_language.md:138]:
```suggestion
requires an unsigned short (two bytes) instead. Please note that the vanilla
```

File: docs/schedule_scripting_language.md [hunk 17]
```diff
@@ -0,0 +1,925 @@
...
+    } else {
+        return_none
+    }
+}
+```
+
+## Syntax
+
+The syntax is simple, it consists on the name of the schedule script followed
+by a sucession of commands. Some commands require arguments (by using
```
Comment (engineer124) [inline docs/schedule_scripting_language.md:216]:
```suggestion
by a succession of commands. Some commands require arguments (by using
```

**Review by engineer124 — APPROVED** [review]
